terraform/node_provider.go

```diff
@@ -1,11 +1,114 @@
 package terraform
 
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/terraform/configs/configschema"
+	"github.com/hashicorp/terraform/providers"
+	"github.com/hashicorp/terraform/tfdiags"
+)
+
 // NodeApplyableProvider represents a provider during an apply.
 type NodeApplyableProvider struct {
 	*NodeAbstractProvider
 }
 
-// GraphNodeEvalable
-func (n *NodeApplyableProvider) EvalTree() EvalNode {
-	return ProviderEvalTree(n, n.ProviderConfig())
+var (
+	_ GraphNodeExecutable = (*NodeApplyableProvider)(nil)
+)
+
+// GraphNodeExecutable
+func (n *NodeApplyableProvider) Execute(ctx EvalContext, op walkOperation) error {
+	_, err := ctx.InitProvider(n.Addr)
+	if err != nil {
+		return err
+	}
+	provider, _, err := GetProvider(ctx, n.Addr)
+	if err != nil {
+		return err
+	}
+
+	switch op {
+	case walkValidate:
+		//Validate
+		return n.ValidateProvider(ctx, provider)
+	case walkRefresh, walkPlan, walkApply, walkDestroy:
+		// EvalConfigProvider
+		return n.ConfigureProvider(ctx, provider, false)
+	case walkImport:
+		return n.ConfigureProvider(ctx, provider, true)
+	}
+	return nil
+}
+
+func (n *NodeApplyableProvider) ValidateProvider(ctx EvalContext, provider providers.Interface) error {
+	var diags tfdiags.Diagnostics
+
+	configBody := buildProviderConfig(ctx, n.Addr, n.ProviderConfig())
+
+	resp := provider.GetSchema()
+	diags = diags.Append(resp.Diagnostics)
+	if diags.HasErrors() {
+		return diags.ErrWithWarnings()
+	}
+
+	configSchema := resp.Provider.Block
+	if configSchema == nil {
+		// Should never happen in real code, but often comes up in tests where
+		// mock schemas are being used that tend to be incomplete.
+		log.Printf("[WARN] ValidateProvider: no config schema is available for %s, so using empty schema", n.Addr)
+		configSchema = &configschema.Block{}
+	}
+
+	configVal, configBody, evalDiags := ctx.EvaluateBlock(configBody, configSchema, nil, EvalDataForNoInstanceKey)
+	diags = diags.Append(evalDiags)
+	if evalDiags.HasErrors() {
+		return diags.ErrWithWarnings()
+	}
+
+	req := providers.PrepareProviderConfigRequest{
+		Config: configVal,
+	}
+
+	validateResp := provider.PrepareProviderConfig(req)
+	diags = diags.Append(validateResp.Diagnostics)
+
+	return diags.ErrWithWarnings()
+}
+
+func (n *NodeApplyableProvider) ConfigureProvider(ctx EvalContext, provider providers.Interface, VerifyConfigIsKnown bool) error {
```

Comment:
With moving this, perhaps tidy up the VerifyConfigIsKnown capitalization? It might also help to have a comment here to explain what that boolean is doing for this